src/render.rs

```diff
@@ -1,145 +1,343 @@
-                chan3byte(buf, PITCH_BEND_CHANGE, c, &v0, &v1)
+            // System common messages
+            MidiMessage::QuarterFrame(value) => {
+                self.write_sys_common_msg(&[QUARTER_FRAME, value.into()])?;
+            }
+            MidiMessage::SongPositionPointer(value) => {
+                let (value_lsb, value_msb) = value.into();
+                self.write_sys_common_msg(&[SONG_POSITION_POINTER, value_lsb, value_msb])?;
             }
-            MidiMessage::SongPositionPointer(v) => {
-                let (v0, v1): (u8, u8) = (*v).into();
-                chan3byte(buf, SONG_POSITION_POINTER, &0, &v0, &v1)
+            MidiMessage::SongSelect(value) => {
+                self.write_sys_common_msg(&[SONG_SELECT, value.into()])?;
             }
-            MidiMessage::ProgramChange(c, p) => chan2byte(buf, PROGRAM_CHANGE, c, p),
-            MidiMessage::ChannelPressure(c, p) => chan2byte(buf, CHANNEL_PRESSURE, c, p),
-            MidiMessage::QuarterFrame(q) => chan2byte(buf, QUARTER_FRAME, &0, q),
-            MidiMessage::SongSelect(s) => chan2byte(buf, SONG_SELECT, &0, s),
-            MidiMessage::TuneRequest => chan1byte(buf, TUNE_REQUEST),
-            MidiMessage::TimingClock => chan1byte(buf, TIMING_CLOCK),
-            MidiMessage::Start => chan1byte(buf, START),
-            MidiMessage::Continue => chan1byte(buf, CONTINUE),
-            MidiMessage::Stop => chan1byte(buf, STOP),
-            MidiMessage::ActiveSensing => chan1byte(buf, ACTIVE_SENSING),
-            MidiMessage::Reset => chan1byte(buf, RESET),
+            MidiMessage::TuneRequest => {
+                self.write_sys_common_msg(&[TUNE_REQUEST])?;
+            }
+
+            // System real time messages
+            MidiMessage::TimingClock => self.transport.write(&[TIMING_CLOCK])?,
+            MidiMessage::Start => self.transport.write(&[START])?,
+            MidiMessage::Continue => self.transport.write(&[CONTINUE])?,
+            MidiMessage::Stop => self.transport.write(&[STOP])?,
+            MidiMessage::ActiveSensing => self.transport.write(&[ACTIVE_SENSING])?,
+            MidiMessage::Reset => self.transport.write(&[RESET])?,
         }
+        Ok(())
+    }
+
+    /// Write a channel voice or channel mode messages, these messages optionally use running status to
+    /// skip sending the status byte
+    fn write_channel_msg(&mut self, data: &[u8]) -> Result<(), T::Error> {
+        let status = data[0];
+        if RUNNING_STATUS && self.running_status == Some(status) {
+            // If the last command written had the same status/channel, the MIDI protocol allows us to
+            // omit sending the status byte again.
+            self.transport.write(&data[1..])?;
+        } else {
+            self.transport.write(data)?;
+        }
+
+        if RUNNING_STATUS {
+            // Store running state so the next message can use it
+            self.running_status = Some(status);
+        }
```

Comment:
technically this `if` could go in the above `else` right? you only need to update the `status` if it hasn't matched?

Reply:
Not always, always need to set the running status in case its the first message of a set with the same status byte. In that case `self.running_status == Some(status)` of the previous if hasn't matched yet.

Reply:
This is what i mean... maybe it is premature optimization but, you already know that `self.running_status == Some(status)` if you get into the top `if` statement, so you only have to set `self.running_status` in the else

```rust
       if RUNNING_STATUS && self.running_status == Some(status) {
            // If the last command written had the same status/channel, the MIDI protocol allows us to
            // omit sending the status byte again.
            self.transport.write(&data[1..])?;
        } else {
            self.transport.write(data)?;

            if RUNNING_STATUS {
                // Store running state so the next message can use it
                self.running_status = Some(status);
            }
        }

```

Reply:
Nope, thats not premature optimization at all. I just had this wrong in my head. This is cleaner. Thanks!